base-containers/base/inginious_container_api/run_student.py

```diff
@@ -221,6 +154,165 @@ def _hack_signals(receive_signal):
+        zmq_socket.send(msgpack.dumps(
+            {"type": "run_student_command", "socket_id": socket_id, "student_container_id": student_container_id, "command": cmd,
+             "teardown_script": teardown_script, "working_dir": working_dir,
+             "ssh": ssh, "user": user, "only_dockers": both_dockers}, use_bin_type=True))
+        zmq_socket.recv() #ignore answer
+        return None
+
+
+def allow_to_send_signals(signal_handler_callback, connection, student_container_id, both_dockers):
+    """ Allow to transfer signals """
+    if signal_handler_callback is not None:
+        if both_dockers:
+            def receive_signal(signum_s):  # send signal directly to student_container
+                signum_data = str(signum_s).zfill(3).encode("utf8")
+                connection.send(signum_data)
+        else:
+            def receive_signal(signum_s):  # send signal to student_container via docker agent
+                signum_data = str(signum_s).zfill(3).encode("utf8")
+                msg = {"type": "student_signal", "student_container_id": student_container_id, "signal_data": signum_data}
+                send_socket = zmq.asyncio.Context().socket(zmq.REQ)
+                send_socket.connect("ipc:///sockets/main.sock")
+                send_socket.send(msgpack.dumps(msg, use_bin_type=True))
+                send_socket.recv()
+        signal_handler_callback(receive_signal)
+
+
+def wait_until_finished(both_dockers, zmq_socket, stdin, stdout, stderr, student_container_id):
+    """ Dynamically handle stdin, stdout and stderr while waiting for final message """
+    # Start a process to handle the stdin and send it to the student_container
+    if not both_dockers:
+        stdin_handler = threading.Thread(target=handle_stdin, args=(stdin, student_container_id), daemon=True)
+        stdin_handler.start()
+
+    # handle the student_container outputs and wait for final message
+    message = None
+    msg_type = None
+    stdout_file = os.fdopen(stdout, 'w')
+    stderr_file = os.fdopen(stderr, 'w')
+
+    while msg_type != "run_student_retval":
+        zmq_socket.send(msgpack.dumps({"type": "run_student_ask_retval"}, use_bin_type=True))  # ping pong socket
+        message = msgpack.loads(zmq_socket.recv(), use_list=False, strict_map_key=False)
+        msg_type = message["type"]
+
+        if msg_type == "stdout":
+            stdout_file.write(message["message"].rstrip()+'\n')
+            stdout_file.flush()
+
+        if msg_type == "stderr":
+            stderr_file.write(message["message"].rstrip()+'\n')
+            stderr_file.flush()
+    return message
+
+
+def handle_stdin(stdin, student_container_id):
+    """ Read the specified stdin and send content to the student container"""
+    my_context = zmq.Context()
+    my_zmq_socket = my_context.socket(zmq.REQ)
+    my_zmq_socket.connect("ipc:///sockets/main.sock")
+    with os.fdopen(stdin, 'r') as stdin_read:
```

Comment:
Indeed here, we are talking about the stdin specified in the run_student arguments (FD). Thus I believe it should be a FD for a file containing text (values / commands etc) or the sys.stdin (0) for example if you want direct interactivity.

After this, the INGInious script (grading_container) send it to the agent (everytime, being from grading to agent or student to agent and vice-versa, the messages on the asyncio stream are always encoded as binaries).

Still, it's true that I considered the FD specified as stdin to be related to a text file or sys.stdin... I am not sure we should consider binaries ?

Reply:
In the best case, you should be able to pipe a binary file through ``run_student``. I am not sure this will work as is because you consider text line delimeters in your code. 

Moreover, you can't really prevent users to pipe binary. In the worst case, the system should NOT break. Using line delimiters to chunk binary data will probably give some funky results.

Reply:
Stdin, Stdout and Stderr are now handled as binary streams (both on grading_container and student_container sides)